File: src/params.py
"""Model parameter definitions.

Defines `~rse_model.params.ModelParams`, a frozen
dataclass containing default constants used throughout the simulation.
"""

from dataclasses import dataclass

@dataclass(frozen=True)
class ModelParams:
    """Parameters for the RSE model.

    Notes
    -----
    Increasing the number of nodes (neural populations) increases spatial
    frequency; however beyond roughly ``N ~ 250`` the forms may not fully develop.

    Attributes
    ----------
    dt : float
        Time step in milliseconds.
    Te : float
        Excitatory time constant (Rule et al. 2011).
    Ti : float
        Inhibitory time constant (Rule et al. 2011).
    Aee, Aei, Aie, Aii : float
        Connection strengths between neural populations (Rule et al. 2011).
    He : float
        Excitatory firing threshold/bias (Rule et al. 2011).
    Hi : float
        Inhibitory firing threshold/bias (Rule et al. 2011).
    Ge : float
        Gain on stimulation / input drive for excitatory population.
    Gi : float
        Gain on stimulation / input drive for inhibitory population.
    Ne : float
        Excitatory noise strength (Rule et al. 2011).
    Ni : float
        Inhibitory noise strength (Rule et al. 2011).
    V : float
        Threshold on the sinusoid controlling the effective duty cycle of the
        stimulus (Rule et al. 2011).
    """

    # Timestep (ms)
    dt: float = 0.2

    # Time constant - Rule et al. 2011
    Te: float = 10.0
    Ti: float = 20.0

    # Connectivity Constants - Rule et al. 2011
    Aee: float = 10
    Aei: float = 12
    Aie: float = 8.5
    Aii: float = 3.0

    # Threshold/Bias (theta) - Rule et al. 2011
    He: float = 2.0
    Hi: float = 3.5

    # Gain on stimulation / Input drive - Rule et al. 2011
    Ge: float = 1
    Gi: float = 0

    # Noise = Rule et al. 2011
    Ne: float = 0.05
    Ni: float = 0.05
    
    # Constant determining proportion of light to dark from Rule et al (2011))
    V: float = 0.8

File: src/Params.jl
struct ModelParams{T<:AbstractFloat}
    dt::T
    Te::T
    Ti::T
    Aee::T
    Aei::T
    Aie::T
    Aii::T
    He::T
    Hi::T
    Ge::T
    Gi::T
    Ne::T
    Ni::T
    V::T
end

function ModelParams{T}(;
    dt=T(0.2),
    Te=T(10.0),
    Ti=T(20.0),
    Aee=T(10.0),
    Aei=T(12.0),
    Aie=T(8.5),
    Aii=T(3.0),
    He=T(2.0),
    Hi=T(3.5),
    Ge=T(1.0),
    Gi=T(0.0),
    Ne=T(0.05),
    Ni=T(0.05),
    V=T(0.8),
) where {T<:AbstractFloat}
    return ModelParams{T}(dt, Te, Ti, Aee, Aei, Aie, Aii, He, Hi, Ge, Gi, Ne, Ni, V)
end

ModelParams(; kwargs...) = ModelParams{Float64}(; kwargs...)
